### PR TITLE
fix: fix the memory leak

### PR DESCRIPTION
**Pull request by ngot**
fix https://github.com/eggjs/egg/issues/3125

### CODE REVIEW

File: lib/io.js [hunk 1]
```diff
@@ -21,38 +23,52 @@ module.exports = app => {
 
   const namespace = config.namespace;
 
-  for (const nsp in namespace) {
-    const connectionMiddlewareConfig = namespace[nsp].connectionMiddleware;
-    const packetMiddlewareConfig = namespace[nsp].packetMiddleware;
+  app.beforeStart(async () => {
```
Comment (ngot) [inline lib/io.js]: It's only possible to get the app. middleware in the beforeStart.

File: lib/io.js [hunk 2]
```diff
@@ -21,38 +23,52 @@ module.exports = app => {
...
+        return mw._name && mw._name.startsWith('session');
+      })[0];
+
+      connectionMiddlewares.unshift(sessionMiddleware);
+      connectionMiddlewares = connectionMiddlewares.map(mw => utils.middleware(mw));
+      connectionMiddlewares = compose(connectionMiddlewares);
+
+      packetMiddlewares.unshift(sessionMiddleware);
+      packetMiddlewares = packetMiddlewares.map(mw => utils.middleware(mw));
+      packetMiddlewares = compose(packetMiddlewares);
```
Comment (ngot) [inline lib/io.js]: All those middlewares operations can be done once when booting.

File: lib/connectionMiddlewareInit.js [hunk 3]
```diff
@@ -2,26 +2,16 @@
 
 const http = require('http');
 const util = require('./util');
-const compose = require('koa-compose');
```
Comment (ngot) [inline lib/connectionMiddlewareInit.js]: There is no need to do the middleware process every connection and every packet event.

**Comment by ngot** [comment]
@kiddnoke Memory leak should be fixed after this PR. Please help me with a double check.

**Comment by kiddnoke** [comment]
@ngot  it's my honor

**Comment by ngot** [comment]
+ egg-socket.io@4.1.5

**Pull request merged**